Marlin/twibus.cpp

```diff
@@ -0,0 +1,134 @@
+#include <Wire.h>
+
+/**
+ * @brief Class constructor
+ * @details Initialized the TWI bus and clears the buffer
+ */
+twibus::twibus() {
+    Wire.begin(); // We use no address so we will join the BUS as the master
+    this->reset();
+}
+
+/**
+ * @brief Reset the buffer
+ * @details Brings the internal buffer to a known-empty state
+ */
+void twibus::reset() {
+    this->addr = 0;
+    this->buffer_s = 0;
+    this->buffer[0] = 0x00;
+}
+
+/**
+ * @brief Sets the target slave address
+ * @details The target slave address is stored so it can be later used when
+ * the complete packet needs to be sent over the bus.
+ *
+ * @param addr 7-bit integer address
+ */
+void twibus::address(uint8_t addr) {
+    this->addr = addr;
+
+    if (marlin_debug_flags & DEBUG_INFO) {
+        SERIAL_ECHOPAIR("twibus::sendto: ", this->addr);
+        SERIAL_EOL;
+    }
+}
+
+/**
+ * @brief Add one byte to the buffer
+ * @details Adds the byte to the buffer in a sequential way, if buffer is full
+ * the request is silently ignored.
+ *
+ * @param c a data byte
+ */
+void twibus::addbyte(char c) {
+    if (buffer_s >= sizeof(this->buffer)) return;
+    this->buffer[this->buffer_s++] = c;
+
+    if (marlin_debug_flags & DEBUG_INFO) {
+        SERIAL_ECHOPAIR("twibus::addbyte: ", this->buffer[this->buffer_s -1]);
+        SERIAL_EOL;
+    }
+}
+
+/**
+ * @brief Send the buffer data to the bus
+ * @details Flushed the buffer into the bus targeting the cached slave device
+ * address.
+ */
+void twibus::send() {
+    if (!this->addr) return;
+    if (marlin_debug_flags & DEBUG_INFO)
+        SERIAL_ECHOLNPGM("twibus::send()");
+
+    Wire.beginTransmission(this->addr);
+    Wire.write(this->buffer, this->buffer_s);
+    Wire.endTransmission();
+
+    // Reset the buffer after sending the data
+    this->reset();
+}
+
+/**
+ * @brief Request data from slave device
+ * @details Requests data from a slave device, when the data is received it will
+ * be relayed to the serial line using a parser-friendly formatting.
+ *
+ * @param bytes the number of bytes to request
+ */
+void twibus::reqbytes(uint8_t bytes) {
+    if (!this->addr) return;
+    if (marlin_debug_flags & DEBUG_INFO) {
+        SERIAL_ECHOPAIR("twibus::reqbytes(): ", bytes);
+        SERIAL_EOL;
+    }
+
+    millis_t t = millis();
+    Wire.requestFrom(this->addr, bytes);
+
+    // requestFrom() is a blocking function
+    while (Wire.available() < bytes) {
+        if (millis() - t >= this->timeout) break;
+        else continue;
+    }
+
+    SERIAL_ECHO_START;
+    SERIAL_ECHOPAIR("I2C-reply: from:", this->addr);
+    SERIAL_ECHOPAIR(" bytes:", Wire.available());
+    SERIAL_ECHOPGM (" data:");
+
+    // Protect against buffer overflows if the number of received bytes
+    // is less than the number of requested bytes
+    uint8_t wba = Wire.available();
+    for (int i = 0; i < wba; i++) SERIAL_CHAR(Wire.read());
+    SERIAL_EOL;
+
+    // Reset the buffer after sending the data
+    this->reset();
+}
```

Comment:
Indentation should be two spaces in these `twibus.*` files. If you add `EXPERIMENTAL_I2CBUS` to either (all) `Configuration.h` or `Configuration_adv.h` files, then I don't mind merging this so users can mess with it (once they know about it, of course).